File: src/test/java/com/dg/watcher/history/graph/HistoryGraphDataSetGeneratorTest.java
package com.dg.watcher.history.graph;

import com.dg.watcher.watching.saving.SizeEntry;
import org.jfree.data.category.DefaultCategoryDataset;
import org.junit.Test;
import java.util.ArrayList;
import static com.dg.watcher.history.graph.HistoryGraphDataSetGenerator.generateGraphDataSet;
import static com.dg.watcher.base.Const.CONVERSION_FACTOR_BYTE_TO_MEGABYTE;
import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;


public class HistoryGraphDataSetGeneratorTest {
    @Test
    public void should_include_all_20_entries_in_the_generated_data_set() {
        // GIVEN
        ArrayList<SizeEntry> entries = new ArrayList<>();
        for(int i = 0; i < 20; ++i) entries.add(new SizeEntry("#" + i, 0L));

        // WHEN
        DefaultCategoryDataset dataSet = generateGraphDataSet(entries);

        // THEN
        assertThat(retrieveEntries(dataSet), is(equalTo(entries)));
    }

    @Test
    public void should_include_the_last_20_entries_in_the_generated_data_set() {
        // GIVEN
        ArrayList<SizeEntry> entries = new ArrayList<>();
        for(int i = 0; i < 35; ++i) entries.add(new SizeEntry("#" + i, 0L));

        // WHEN
        DefaultCategoryDataset dataSet = generateGraphDataSet(entries);

        // THEN
        assertThat(retrieveEntries(dataSet), is(equalTo(entries.subList(15, 35))));
    }

    @Test
    public void should_use_a_entries_size_in_megabyte_as_its_row_value() {
        // GIVEN
        ArrayList<SizeEntry> entries = new ArrayList<>();
        entries.add(new SizeEntry("#1", 10000000L));

        // WHEN
        DefaultCategoryDataset dataSet = generateGraphDataSet(entries);

        // THEN
        float rowValue = dataSet.getValue(0, 0).floatValue();

        assertThat(rowValue, is(10000000L / CONVERSION_FACTOR_BYTE_TO_MEGABYTE));
    }

    @Test
    public void should_use_a_entries_build_name_as_its_column_key() {
        // GIVEN
        ArrayList<SizeEntry> entries = new ArrayList<>();
        entries.add(new SizeEntry("#1", 10000000L));

        // WHEN
        DefaultCategoryDataset dataSet = generateGraphDataSet(entries);

        // THEN
        String columnKey = dataSet.getColumnKey(0).toString();

        assertThat(columnKey, is(equalTo("#1")));
    }

    private ArrayList<SizeEntry> retrieveEntries(DefaultCategoryDataset dataSet) {
        ArrayList<SizeEntry> entries = new ArrayList<>();

        for(int i = 0; i < dataSet.getColumnCount(); ++i) {
            String buildName = dataSet.getColumnKey(i).toString();

            float apkSizeInMb = dataSet.getValue(0, i).floatValue();
            long  apkSizeInByte = (long) (apkSizeInMb
                    * CONVERSION_FACTOR_BYTE_TO_MEGABYTE);

            entries.add(new SizeEntry(buildName, apkSizeInByte));
        }

        return entries;
    }
}

File: src/test/kotlin/com/dg/watcher/history/graph/HistoryGraphDataSetGeneratorTest.kt
package com.dg.watcher.history.graph

import com.dg.watcher.base.CONVERSION_FACTOR_BYTE_TO_MEGABYTE
import com.dg.watcher.watching.saving.SizeEntry
import org.hamcrest.Matchers.`is`
import org.hamcrest.Matchers.equalTo
import org.jfree.data.category.DefaultCategoryDataset
import org.junit.Assert.assertThat
import org.junit.Test


class HistoryGraphDataSetGeneratorTest {
    @Test
    fun `Should include all 20 entries in the generated data set`() {
        // GIVEN
        val entries = arrayListOf<SizeEntry>()
        repeat(20) { entries.add(SizeEntry("#" + it, 0L)) }

        // WHEN
        val dataSet = generateGraphDataSet(entries)

        // THEN
        assertThat(retrieveEntries(dataSet), `is`(equalTo(entries)))
    }

    @Test
    fun `Should include the last 20 entries in the generated data set`() {
        // GIVEN
        val entries = arrayListOf<SizeEntry>()
        repeat(35) { entries.add(SizeEntry("#" + it, 0L)) }

        // WHEN
        val dataSet = generateGraphDataSet(entries)

        // THEN
        assertThat(retrieveEntries(dataSet), `is`(equalTo(entries.subList(15, 35))))
    }

    @Test
    fun `Should convert a entries size to megabyte for its row value`() {
        // GIVEN
        val entries = arrayListOf(SizeEntry("#1", sizeInByte = 10000000L))

        // WHEN
        val dataSet = generateGraphDataSet(entries)

        // THEN
        val rowValue = dataSet.getValue(0, 0).toFloat()

        assertThat(rowValue, `is`(10000000L / CONVERSION_FACTOR_BYTE_TO_MEGABYTE))
    }

    @Test
    fun `Should use a entries build name as its column key`() {
        // GIVEN
        val entries = arrayListOf(SizeEntry("#1", 0L))

        // WHEN
        val dataSet = generateGraphDataSet(entries)

        // THEN
        val columnKey = dataSet.getColumnKey(0).toString()

        assertThat(columnKey, `is`(equalTo("#1")))
    }

    private fun retrieveEntries(dataSet: DefaultCategoryDataset) = arrayListOf<SizeEntry>().apply {
        for(i in 0 until dataSet.columnCount) {
            val buildName = dataSet.getColumnKey(i).toString()

            val apkSizeInMb = dataSet.getValue(0, i).toFloat()
            val apkSizeInByte = (apkSizeInMb * CONVERSION_FACTOR_BYTE_TO_MEGABYTE).toLong()

            add(SizeEntry(buildName, apkSizeInByte))
        }
    }
}